snowfakery/standard_plugins/datasets.py

```diff
@@ -0,0 +1,220 @@
+    #     copying the rows up-front is better)
+    #
+    #   * segment the file into hundred-thousand-row partitions. Shuffle the
+    #     rows in each partition and then pick randomly among the partitions
+    #     before grabbing a row
+    def start(self):
+        self.file.seek(0)
+        d = DictReader(self.file)
+        rows = [PluginResult(row) for row in d]
+        shuffle(rows)
+
+        self.results = iter(rows)
+
+
+class Dataset(SnowfakeryPlugin):
+    class Functions:
+
+        datasets = {}
+
+        def iterate(self, *args, **kwargs):
+            return self._iterate(*args, **kwargs, iteration_mode="linear")
+
+        def shuffle(self, *args, **kwargs):
+            return self._iterate(*args, **kwargs, iteration_mode="shuffle")
+
+        def _iterate(self, dataset, table=None, name=None, iteration_mode="linear"):
+            tablename = table
+            name = name or self.context.field_vars()["template"].id
+            key = (dataset, tablename, name)
+            dataset_instance = self.datasets.get(key)
+            if not dataset_instance:
+                filename = self.context.field_vars()["template"].filename
+                assert filename
+                rootpath = Path(filename).parent
+                dataset_instance = self.datasets[key] = self._load_dataset(
+                    dataset, tablename, rootpath, iteration_mode
+                )
+            rc = next(dataset_instance, None)
+            if not rc:
+                dataset_instance.start()
+                rc = next(dataset_instance, None)
+
+            assert rc is not None
+            return rc
+
+        def _load_dataset(self, dataset, tablename, rootpath, iteration_mode):
+            with chdir(rootpath):
+                if "://" in dataset:
+                    return sql_dataset(dataset, tablename, iteration_mode)
+                else:
+                    filename = Path(dataset)
+
+                    if not filename.exists():
+                        raise FileNotFoundError("File not found:" + str(filename))
+
+                    if filename.suffix != ".csv":
+                        raise AssertionError(
+                            f"Filename extension must be .csv, not {filename.suffix}"
+                        )
+
+                    if iteration_mode == "linear":
+                        return CSVDatasetLinearIterator(filename)
+                    elif iteration_mode == "shuffle":
+                        return CSVDatasetRandomPermutationIterator(filename)
+
+        def close(self):
+            for dataset in self.datasets.values():
+                dataset.close()
+
+
+@contextmanager
+def chdir(path):
+    """Context manager that changes to another directory
+
+    Not thread-safe!!!
+    """
+    cwd = os.getcwd()
+    os.chdir(path)
+    try:
+        yield
+    finally:
+        os.chdir(cwd)
+
+
+yaml.SafeDumper.add_representer(quoted_name, Representer.represent_str)
```

Comment:
This seems like a potentially surprising side effect of importing this plugin. It would be cleaner if snowfakery used its own dumper, perhaps.

Reply:
Okay I'll add an issue for this. quoted_name objects ARE strings (subclassing) with an extra boolean which seems to me to be relevant only in exceedingly rare situations so I wouldn't call this a blocker. But sure I can fix it up in the next month or so in case Snowfakery is ever embedded in some giant system that round-trips weird tablenames through YAML and has tablenames that differ only in case(?). I agree that it is more polite to try and avoid such a situation even if it seems unlikely.

Reply:
https://github.com/SFDO-Tooling/Snowfakery/issues/132